### ISSUE TITLE
ERROR: Error when fetching backup: pg_basebackup exited with code=1

### ISSUE THREAD

**Issue by sandeepkalra** [zalando/patroni#1350]
Hi, Can someone please guide me on this. I am configuring Postgres 12.1 with Patroni. I have a cluster with 7 nodes.Everytime I scale up, i end up with this situation. The master/Leader starts and works but the slave/replica either stops or ends up in 'creating replica' mode. 
```
root@103d123f2d5c:/bp2/src# patronictl -c pg_patroni.yml list
+---------+--------+----------------+--------+------------------+----+-----------+
| Cluster | Member |      Host      |  Role  |      State       | TL | Lag in MB |
+---------+--------+----------------+--------+------------------+----+-----------+
|  blue0  |  pg_0  | 127.0.0.1:5432 | Leader |     running      |  1 |           |
|  blue0  |  pg_1  | 127.0.0.1:5432 |        |     stopped      |    |   unknown |
|  blue0  |  pg_2  | 127.0.0.1:5432 |        |     stopped      |    |   unknown |
|  blue0  |  pg_3  | 127.0.0.1:5432 |        |     stopped      |    |   unknown |
|  blue0  |  pg_4  | 127.0.0.1:5432 |        |     stopped      |    |   unknown |
|  blue0  |  pg_5  | 127.0.0.1:5432 |        |     stopped      |    |   unknown |
|  blue0  |  pg_6  | 127.0.0.1:5432 |        | creating replica |    |   unknown |
+---------+--------+----------------+--------+------------------+----+-----------+

The logs constantly shows error such as these: 

2020-01-02 23:04:27,006 DEBUG: Sending request(xid=717): SetData(path='/bp/blue0/members/pg_1', data=b'{"conn_url":"postgres://127.0.0.1:5432/postgres","api_url":"http://127.0.0.1:8008/patroni","state":"stopped","role":"uninitialized","version":"1.6.3"}', version=-1)
2020-01-02 23:04:27,011 DEBUG: Received response(xid=717): ZnodeStat(czxid=12885154300, mzxid=12885161237, ctime=1577999766021, mtime=1578006267006, version=652, cversion=0, aversion=0, ephemeralOwner=31334655534829047, dataLength=150, numChildren=0, pzxid=12885154300)
2020-01-02 23:04:27,011 INFO: trying to bootstrap from leader 'pg_0'
2020-01-02 23:04:27,025 ERROR: Error when fetching backup: pg_basebackup exited with code=1
2020-01-02 23:04:27,025 WARNING: Trying again in 5 seconds
2020-01-02 23:04:32,037 ERROR: Error when fetching backup: pg_basebackup exited with code=1
2020-01-02 23:04:32,037 ERROR: failed to bootstrap from leader 'pg_0'
2020-01-02 23:04:32,037 INFO: Removing data directory: /bp2/data/psql
2020-01-02 23:04:37,004 INFO: Lock owner: pg_0; I am pg_1
2020-01-02 23:04:37,006 DEBUG: Sending request(xid=718): SetData(path='/bp/blue0/members/pg_1', data=b'{"conn_url":"postgres://127.0.0.1:5432/postgres","api_url":"http://127.0.0.1:8008/patroni","state":"stopped","role":"uninitialized","version":"1.6.3"}', version=-1)
2020-01-02 23:04:37,011 DEBUG: Received response(xid=718): ZnodeStat(czxid=12885154300, mzxid=12885161248, ctime=1577999766021, mtime=1578006277007, version=653, cversion=0, aversion=0, ephemeralOwner=31334655534829047, dataLength=150, numChildren=0, pzxid=12885154300)
2020-01-02 23:04:37,012 INFO: trying to bootstrap from leader 'pg_0'
2020-01-02 23:04:37,022 ERROR: Error when fetching backup: pg_basebackup exited with code=1
2020-01-02 23:04:37,023 WARNING: Trying again in 5 seconds
2020-01-02 23:04:42,036 ERROR: Error when fetching backup: pg_basebackup exited with code=1
2020-01-02 23:04:42,036 ERROR: failed to bootstrap from leader 'pg_0'
2020-01-02 23:04:42,036 INFO: Removing data directory: /bp2/data/psql

```

and here is my yaml file:
```
scope: blue0
namespace: /bp/
name: pg_1

log:
  level: DEBUG
  traceback_level: debug
  dir: /bp2/log/
   
restapi:
  listen: 127.0.0.1:8008
  connect_address: 127.0.0.1:8008

zookeeper:
  hosts: zookeeper:2181

bootstrap:
  dcs:
    ttl: 30
    loop_wait: 10
    retry_timeout: 10
    maximum_lag_on_failover: 1048576
    postgresql:
      use_pg_rewind: true
      parameters:
  initdb: 
    - encoding: UTF8
    - data-checksums
  pg_hba: 
    - host replication replicator 127.0.0.1/32 md5
    - host all all 0.0.0.0/0 md5
  users:
    sbpadmin:
        password: sbpadminpw
        options:
            - createrole
            - createdb
    bpadmin:
        password: bpadminpw
        options:
            - replication
    wbpadmin:
        password: wbpadminpw
        options:
            - rewind

postgresql:
  listen: 127.0.0.1:5432
  connect_address: 127.0.0.1:5432
  config_dir: /bp2/data/psql
  data_dir: /bp2/data/psql
  bin_dir: /usr/lib/postgresql/12/bin/

  pgpass: /bp2/log/pgpass
  authentication:
    replication:
      username: bpadmin
      password: bpadminpw
    superuser:
      username: sbpadmin
      password: sbpadminpw
    rewind:
      username: wbpadmin
      password: wbpadminpw
  parameters:
    unix_socket_directories: '.'
tags:
    nofailover: false
    noloadbalance: false
    clonefrom: false
    nosync: false
```

**Comment by CyberDem0n** [comment]
Why do you set listen and connect_address to 127.0.0.1?
How nodes will discover each other?

**Comment by sandeepkalra** [comment]
I have corrected the IP, still the problem persists:
```
root@ac1e1023ce87:/bp2/src# patronictl -c pg_patroni.yml list
+--------------+--------+-------------+--------+------------------+----+-----------+
|   Cluster    | Member |     Host    |  Role  |      State       | TL | Lag in MB |
+--------------+--------+-------------+--------+------------------+----+-----------+
| bluecluster0 |  pg_0  | 172.16.0.50 | Leader |     running      |  1 |           |
| bluecluster0 |  pg_1  | 172.16.0.53 |        | creating replica |    |   unknown |
| bluecluster0 |  pg_2  | 172.16.0.59 |        |     stopped      |    |   unknown |
+--------------+--------+-------------+--------+------------------+----+-----------+
```

This is the capture of logs from pg_1
```
2020-01-06 04:38:33,962 DEBUG: Sending request(xid=18): GetData(path='/bpspace/bluecluster0/leader', watcher=None)
2020-01-06 04:38:33,962 DEBUG: Received response(xid=18): (b'pg_0', ZnodeStat(czxid=12885238553, mzxid=12885238553, ctime=1578285488163, mtime=1578285488163, version=0, cversion=0, aversion=0, ephemeralOwner=103392249575575758, dataLength=4, numChildren=0, pzxid=12885238553))
2020-01-06 04:38:33,963 INFO: Lock owner: pg_0; I am pg_1
2020-01-06 04:38:33,963 DEBUG: Sending request(xid=19): SetData(path='/bpspace/bluecluster0/members/pg_1', data=b'{"conn_url":"postgres://172.16.0.53:5432/postgres","api_url":"http://127.0.0.1:8008/patroni","state":"creating replica","role":"uninitialized","version":"1.6.3"}', version=-1)
2020-01-06 04:38:33,965 ERROR: Error when fetching backup: pg_basebackup exited with code=1
2020-01-06 04:38:33,965 WARNING: Trying again in 5 seconds
2020-01-06 04:38:33,967 DEBUG: Received response(xid=19): ZnodeStat(czxid=12885238573, mzxid=12885238576, ctime=1578285513939, mtime=1578285513963, version=2, cversion=0, aversion=0, ephemeralOwner=103392249575575760, dataLength=161, numChildren=0, pzxid=12885238573)
2020-01-06 04:38:33,967 INFO: bootstrap from leader 'pg_0' in progress
2020-01-06 04:38:38,982 ERROR: Error when fetching backup: pg_basebackup exited with code=1
2020-01-06 04:38:38,983 ERROR: failed to bootstrap from leader 'pg_0'
2020-01-06 04:38:38,983 INFO: Removing data directory: /bp2/data/psql
2020-01-06 04:38:43,948 INFO: Lock owner: pg_0; I am pg_1
2020-01-06 04:38:43,949 INFO: trying to bootstrap from leader 'pg_0'
2020-01-06 04:38:43,960 ERROR: Error when fetching backup: pg_basebackup exited with code=1
2020-01-06 04:38:43,961 WARNING: Trying again in 5 seconds
2020-01-06 04:38:48,974 ERROR: Error when fetching backup: pg_basebackup exited with code=1
2020-01-06 04:38:48,974 ERROR: failed to bootstrap from leader 'pg_0'
2020-01-06 04:38:48,974 INFO: Removing data directory: /bp2/data/psql
2020-01-06 04:38:53,948 INFO: Lock owner: pg_0; I am pg_1
2020-01-06 04:38:53,949 INFO: trying to bootstrap from leader 'pg_0'
2020-01-06 04:38:53,960 ERROR: Error when fetching backup: pg_basebackup exited with code=1
2020-01-06 04:38:53,960 WARNING: Trying again in 5 seconds
2020-01-06 04:38:58,971 ERROR: Error when fetching backup: pg_basebackup exited with code=1
2020-01-06 04:38:58,971 ERROR: failed to bootstrap from leader 'pg_0'
2020-01-06 04:38:58,972 INFO: Removing data directory: /bp2/data/psql
2020-01-06 04:39:03,948 INFO: Lock owner: pg_0; I am pg_1
2020-01-06 04:39:03,949 INFO: trying to bootstrap from leader 'pg_0'
2020-01-06 04:39:03,960 ERROR: Error when fetching backup: pg_basebackup exited with code=1
2020-01-06 04:39:03,961 WARNING: Trying again in 5 seconds
2020-01-06 04:39:08,974 ERROR: Error when fetching backup: pg_basebackup exited with code=1
2020-01-06 04:39:08,974 ERROR: failed to bootstrap from leader 'pg_0'
2020-01-06 04:39:08,974 INFO: Removing data directory: /bp2/data/psql
2020-01-06 04:39:13,948 INFO: Lock owner: pg_0; I am pg_1
2020-01-06 04:39:13,949 INFO: trying to bootstrap from leader 'pg_0'
2020-01-06 04:39:13,959 ERROR: Error when fetching backup: pg_basebackup exited with code=1
2020-01-06 04:39:13,959 WARNING: Trying again in 5 seconds
2020-01-06 04:39:18,973 ERROR: Error when fetching backup: pg_basebackup exited with code=1
2020-01-06 04:39:18,973 ERROR: failed to bootstrap from leader 'pg_0'
2020-01-06 04:39:18,973 INFO: Removing data directory: /bp2/data/psql
2020-01-06 04:39:23,948 INFO: Lock owner: pg_0; I am pg_1
2020-01-06 04:39:23,949 INFO: trying to bootstrap from leader 'pg_0'
2020-01-06 04:39:23,959 ERROR: Error when fetching backup: pg_basebackup exited with code=1
2020-01-06 04:39:23,959 WARNING: Trying again in 5 seconds
2020-01-06 04:39:28,974 ERROR: Error when fetching backup: pg_basebackup exited with code=1
2020-01-06 04:39:28,974 ERROR: failed to bootstrap from leader 'pg_0'
2020-01-06 04:39:28,974 INFO: Removing data directory: /bp2/data/psql
2020-01-06 04:39:33,953 INFO: Lock owner: pg_0; I am pg_1
2020-01-06 04:39:33,953 INFO: trying to bootstrap from leader 'pg_0'
2020-01-06 04:39:33,965 ERROR: Error when fetching backup: pg_basebackup exited with code=1
2020-01-06 04:39:33,965 WARNING: Trying again in 5 seconds
```

**Comment by CyberDem0n** [comment]
I have no idea how your new config looks like, but I can see from logs that `restapi.connect_address` is still set to the localhost! This is absolutely wrong, please read the documentation: https://patroni.readthedocs.io/en/latest/SETTINGS.html#rest-api

Also I am not sure that you changed `postgresql.listen_address`.
In any case please carefully examine logs, especially postgres logs.

**Comment by sandeepkalra** [comment]
Yes, I did the changes in steps, and it could be that when i captured the logs then restapi.connect_address may be 127.0.0.1. 
Here is the full yaml file, and the status of all nodes again:
```
root@fae1466baba3:/bp2/src# patronictl -c pg_patroni.yml list
+--------------+--------+-------------+--------+------------------+----+-----------+
|   Cluster    | Member |     Host    |  Role  |      State       | TL | Lag in MB |
+--------------+--------+-------------+--------+------------------+----+-----------+
| bluecluster0 |  pg_0  | 172.16.0.50 | Leader |     running      |  1 |           |
| bluecluster0 |  pg_1  | 172.16.0.53 |        |     stopped      |    |   unknown |
| bluecluster0 |  pg_2  | 172.16.0.59 |        |     stopped      |    |   unknown |
| bluecluster0 |  pg_3  | 172.16.0.60 |        |     stopped      |    |   unknown |
| bluecluster0 |  pg_4  | 172.16.0.61 |        |     stopped      |    |   unknown |
| bluecluster0 |  pg_5  | 172.16.0.62 |        | creating replica |    |   unknown |
| bluecluster0 |  pg_6  | 172.16.0.63 |        |     stopped      |    |   unknown |
+--------------+--------+-------------+--------+------------------+----+-----------+



root@fae1466baba3:/bp2/src# cat pg_patroni.yml 
scope: bluecluster0
namespace: /bpspace/
name: pg_3

log:
  level: DEBUG
  traceback_level: debug
  dir: /bp2/log/
   
restapi:
  listen: 172.16.0.60:8008
  connect_address: 172.16.0.60:8008

zookeeper:
  hosts: zookeeper:2181

bootstrap:
  dcs:
    ttl: 30
    loop_wait: 10
    retry_timeout: 10
    maximum_lag_on_failover: 1048576
    postgresql:
      use_pg_rewind: true
      parameters:
  initdb: 
    - encoding: UTF8
    - data-checksums
  pg_hba: 
    - host replication replicator 127.0.0.1/32 md5
    - host all all 0.0.0.0/0 md5
  users:
    sbpadmin:
        password: sbpadminpw
        options:
            - createrole
            - createdb
    bpadmin:
        password: bpadminpw
        options:
            - replication
    wbpadmin:
        password: wbpadminpw
        options:
            - rewind

postgresql:
  listen: 172.16.0.60:5432
  connect_address: 172.16.0.60:5432
  config_dir: /bp2/data/psql
  data_dir: /bp2/data/psql
  bin_dir: /usr/lib/postgresql/12/bin/

  pgpass: /bp2/log/pgpass
  authentication:
    replication:
      username: bpadmin
      password: bpadminpw
    superuser:
      username: sbpadmin
      password: sbpadminpw
    rewind:
      username: wbpadmin
      password: wbpadminpw
  parameters:
    unix_socket_directories: '.'
tags:
    nofailover: false
    noloadbalance: false
    clonefrom: false
    nosync: false
```

**Comment by CyberDem0n** [comment]
Have you checked that nodes can access each other via port 5432?

**Comment by sandeepkalra** [comment]
I can ping the IP address, but cant find the port active. Patroni shows "stopped"  or "creating replica" status for those nodes, and I am not sure if port  5432 will be reachable before the status turn to "running"


```
root@3fa018ff3bdb:/bp2/src# patronictl -c pg_patroni.yml list
+---------+--------+-------------+--------+------------------+----+-----------+
| Cluster | Member |     Host    |  Role  |      State       | TL | Lag in MB |
+---------+--------+-------------+--------+------------------+----+-----------+
|    c4   |  pg_0  | 172.16.0.50 | Leader |     running      |  1 |           |
|    c4   |  pg_1  | 172.16.0.53 |        |     stopped      |    |   unknown |
|    c4   |  pg_2  | 172.16.0.59 |        |     stopped      |    |   unknown |
|    c4   |  pg_3  | 172.16.0.60 |        |     stopped      |    |   unknown |
|    c4   |  pg_4  | 172.16.0.61 |        |     stopped      |    |   unknown |
|    c4   |  pg_5  | 172.16.0.62 |        | creating replica |    |   unknown |
|    c4   |  pg_6  | 172.16.0.63 |        | creating replica |    |   unknown |
+---------+--------+-------------+--------+------------------+----+-----------+
root@3fa018ff3bdb:/bp2/src# nmap -sS 172.16.0.63 -p 5432

Starting Nmap 7.60 ( https://nmap.org ) at 2020-01-08 20:02 UTC
Nmap scan report for 3fa018ff3bdb (172.16.0.63)
Host is up (0.000081s latency).

PORT     STATE  SERVICE
5432/tcp closed postgresql

Nmap done: 1 IP address (1 host up) scanned in 0.31 seconds

root@3fa018ff3bdb:/bp2/src# ping 172.16.0.63
PING 172.16.0.63 (172.16.0.63) 56(84) bytes of data.
64 bytes from 172.16.0.63: icmp_seq=1 ttl=64 time=0.037 ms
64 bytes from 172.16.0.63: icmp_seq=2 ttl=64 time=0.042 ms
^C
--- 172.16.0.63 ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 999ms
rtt min/avg/max/mdev = 0.037/0.039/0.042/0.006 ms


````

**Comment by CyberDem0n** [comment]
Every node MUST have access to any other node via port 5432. Have you checked that?

**Comment by sandeepkalra** [comment]
Yes, there is access to port 5432 from each cluster node.

**Comment by CyberDem0n** [comment]
Magic doesn't exists. pg_basebackup can't fetch basebackup from the primary, you have to figure out why.

**Comment by sandeepkalra** [comment]
Thanks. I think I found the issue, but do not know if my fix is correct or not. There were 2 problems 
1. There was no user called 'replicator'. The pg_hba entry wasn't correct. I have changed hba block to
```
pg_hba:
 host replication $rep_username $localhost_ip/32 md5
 host all   $rep_username $localhost_ip/32 md5
 host replication $rep_username 0.0.0.0/0 md5
 host all   $rep_username 0.0.0.0/0 md5
 host all all 0.0.0.0/0 md5
```

the $rep_username and $localhost_ip were replaced with the correct username and host-ip.

After this, the replication completes from the replicas, but still, they end up an error due to `missing postmaster.opts` ending in 'start failed' state. The 'reinit' on replicas worked and restarts the node. I think the pg_hba still isn't correct for some reason. 

I will close this help ticket soon but it will be great if someone can suggest to me what may be the reason for missing the postmaster.opts  (this file exists on master). 

I have one more followup question. The pg_hba section for each cluster node has its own host-ip. I have an understanding that Patroni will find leader using zookeeper and then know which leader to connect but I have seen examples where the leader pg_hba has an entry for all slave nodes. Is that necessary?


Thanks,
